// jupyterlab/lib/launcher/index.js
// Copyright (c) Jupyter Development Team.
// Distributed under the terms of the Modified BSD License.
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var iteration_1 = require('phosphor/lib/algorithm/iteration');
var vector_1 = require('phosphor/lib/collections/vector');
var disposable_1 = require('phosphor/lib/core/disposable');
var token_1 = require('phosphor/lib/core/token');
var vdom_1 = require('phosphor/lib/ui/vdom');
var vdom_2 = require('../common/vdom');
/* tslint:disable */
/**
 * The launcher token.
 */
exports.ILauncher = new token_1.Token('jupyter.services.launcher');
/* tslint:enable */
/**
 * The class name added to LauncherWidget instances.
 */
var LAUNCHER_CLASS = 'jp-LauncherWidget';
/**
 * The class name added to LauncherWidget image nodes.
 */
var IMAGE_CLASS = 'jp-LauncherWidget-image';
/**
 * The class name added to LauncherWidget text nodes.
 */
var TEXT_CLASS = 'jp-LauncherWidget-text';
/**
 * The class name added to LauncherWidget item nodes.
 */
var ITEM_CLASS = 'jp-LauncherWidget-item';
/**
 * The class name added to LauncherWidget folder node.
 */
var FOLDER_CLASS = 'jp-LauncherWidget-folder';
/**
 * The class name added for the folder icon from default-theme.
 */
var FOLDER_ICON_CLASS = 'jp-FolderIcon';
/**
 * The class name added to LauncherWidget path nodes.
 */
var PATH_CLASS = 'jp-LauncherWidget-path';
/**
 * The class name added to LauncherWidget current working directory node.
 */
var CWD_CLASS = 'jp-LauncherWidget-cwd';
/**
 * The class name added to LauncherWidget body nodes.
 */
var BODY_CLASS = 'jp-LauncherWidget-body';
/**
 * The class name added to LauncherWidget dialog node.
 */
var DIALOG_CLASS = 'jp-LauncherWidget-dialog';
/**
 * LauncherModel keeps track of the path to working directory and has a list of
 * LauncherItems, which the LauncherWidget will render.
 */
var LauncherModel = (function (_super) {
    __extends(LauncherModel, _super);
    /**
     * Create a new launcher model.
     */
    function LauncherModel() {
        _super.call(this);
        this._items = null;
        this._path = 'home';
        this._items = new vector_1.Vector();
    }
    Object.defineProperty(LauncherModel.prototype, "path", {
        /**
         * The path to the current working directory.
         */
        get: function () {
            return this._path;
        },
        set: function (path) {
            if (path === this._path) {
                return;
            }
            this._path = path;
            this.stateChanged.emit(void 0);
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Add a command item to the launcher, and trigger re-render event for parent
     * widget.
     *
     * @param options - The specification options for a launcher item.
     *
     * @returns A disposable that will remove the item from Launcher, and trigger
     * re-render event for parent widget.
     *
     */
    LauncherModel.prototype.add = function (options) {
        var _this = this;
        // Create a copy of the options to circumvent mutations to the original.
        var item = JSON.parse(JSON.stringify(options));
        // If image class name is not set, use the default value.
        item.imgClassName = item.imgClassName ||
            "jp-Image" + item.name.replace(/\ /g, '');
        this._items.pushBack(item);
        this.stateChanged.emit(void 0);
        return new disposable_1.DisposableDelegate(function () {
            _this._items.remove(item);
            _this.stateChanged.emit(void 0);
        });
    };
    /**
     * Return an iterator of launcher items.
     */
    LauncherModel.prototype.items = function () {
        return this._items.iter();
    };
    return LauncherModel;
}(vdom_2.VDomModel));
exports.LauncherModel = LauncherModel;
/**
 * A virtual-DOM-based widget for the Launcher.
 */
var LauncherWidget = (function (_super) {
    __extends(LauncherWidget, _super);
    /**
     * Construct a new launcher widget.
     */
    function LauncherWidget(options) {
        _super.call(this);
        this._linker = null;
        this.addClass(LAUNCHER_CLASS);
        this._linker = options.linker;
    }
    /**
     * Render the launcher to virtual DOM nodes.
     */
    LauncherWidget.prototype.render = function () {
        var _this = this;
        // Create an iterator that yields rendered item nodes.
        var children = iteration_1.map(this.model.items(), function (item) {
            var img = vdom_1.h.span({ className: item.imgClassName + ' ' + IMAGE_CLASS });
            var text = vdom_1.h.span({ className: TEXT_CLASS }, item.name);
            var attrs = _this._linker.populateVNodeAttrs({
                className: ITEM_CLASS
            }, item.command, item.args);
            return vdom_1.h.div(attrs, [img, text]);
        });
        var folderImage = vdom_1.h.span({
            className: FOLDER_CLASS + " " + FOLDER_ICON_CLASS
        });
        var p = this.model.path;
        var pathName = p.length ? "home > " + p.replace(/\//g, ' > ') : 'home';
        var path = vdom_1.h.span({ className: PATH_CLASS }, pathName);
        var cwd = vdom_1.h.div({ className: CWD_CLASS }, [folderImage, path]);
        var body = vdom_1.h.div({ className: BODY_CLASS }, iteration_1.toArray(children));
        return vdom_1.h.div({ className: DIALOG_CLASS }, [cwd, body]);
    };
    return LauncherWidget;
}(vdom_2.VDomWidget));
exports.LauncherWidget = LauncherWidget;
