// jupyterlab/lib/mainmenu/plugin.js
// Copyright (c) Jupyter Development Team.
// Distributed under the terms of the Modified BSD License.
"use strict";
var widget_1 = require('phosphor/lib/ui/widget');
var _1 = require('./');
/**
 * The class name for all main area portrait tab icons.
 */
var PORTRAIT_ICON_CLASS = 'jp-MainAreaPortraitIcon';
/**
 * The class name for the jupyter icon from the default theme.
 */
var JUPYTER_ICON_CLASS = 'jp-JupyterIcon';
/**
 * A service providing an interface to the main menu.
 */
exports.mainMenuProvider = {
    id: 'jupyter.services.main-menu',
    provides: _1.IMainMenu,
    activate: activateMainMenu
};
/**
 * Activate the main menu extension.
 */
function activateMainMenu(app) {
    var menu = new _1.MainMenu({ keymap: app.keymap });
    menu.id = 'jp-MainMenu';
    var logo = new widget_1.Widget();
    logo.node.className = PORTRAIT_ICON_CLASS + " " + JUPYTER_ICON_CLASS;
    logo.id = 'jp-MainLogo';
    app.shell.addToTopArea(logo);
    app.shell.addToTopArea(menu);
    return menu;
}
