// jupyterlab/lib/notebook/notebook/widgetfactory.js
// Copyright (c) Jupyter Development Team.
// Distributed under the terms of the Modified BSD License.
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var docregistry_1 = require('../../docregistry');
var default_toolbar_1 = require('./default-toolbar');
var panel_1 = require('./panel');
/**
 * A widget factory for notebook panels.
 */
var NotebookWidgetFactory = (function (_super) {
    __extends(NotebookWidgetFactory, _super);
    /**
     * Construct a new notebook widget factory.
     *
     * @param options - The options used to construct the factory.
     */
    function NotebookWidgetFactory(options) {
        _super.call(this, options);
        this._rendermime = null;
        this._clipboard = null;
        this._renderer = null;
        this._rendermime = options.rendermime;
        this._clipboard = options.clipboard;
        this._renderer = options.renderer;
    }
    /**
     * Dispose of the resources used by the factory.
     */
    NotebookWidgetFactory.prototype.dispose = function () {
        if (this.isDisposed) {
            return;
        }
        this._rendermime = null;
        this._clipboard = null;
        _super.prototype.dispose.call(this);
    };
    /**
     * Create a new widget.
     *
     * #### Notes
     * The factory will start the appropriate kernel and populate
     * the default toolbar items using `ToolbarItems.populateDefaults`.
     */
    NotebookWidgetFactory.prototype.createNewWidget = function (context) {
        var rendermime = this._rendermime.clone();
        var panel = new panel_1.NotebookPanel({
            rendermime: rendermime,
            clipboard: this._clipboard,
            renderer: this._renderer
        });
        panel.context = context;
        default_toolbar_1.ToolbarItems.populateDefaults(panel);
        return panel;
    };
    return NotebookWidgetFactory;
}(docregistry_1.ABCWidgetFactory));
exports.NotebookWidgetFactory = NotebookWidgetFactory;
