// jupyterlab/lib/main/plugin.js
// Copyright (c) Jupyter Development Team.
// Distributed under the terms of the Modified BSD License.
"use strict";
var commandpalette_1 = require('../commandpalette');
/**
 * The main extension.
 */
exports.mainExtension = {
    id: 'jupyter.extensions.main',
    requires: [commandpalette_1.ICommandPalette],
    activate: function (app, palette) {
        var commandId = 'main-jupyterlab:closeAll';
        app.commands.addCommand(commandId, {
            label: 'Close All Widgets',
            execute: function () {
                app.shell.closeAll();
            }
        });
        palette.addItem({ command: commandId, category: 'Main Area' });
        var message = 'Are you sure you want to exit JupyterLab?\n' +
            'Any unsaved changes will be lost.';
        // The spec for the `beforeunload` event is implemented differently by
        // the different browser vendors. Consequently, the `event.returnValue`
        // attribute needs to set in addition to a return value being returned.
        // For more information, see:
        // https://developer.mozilla.org/en/docs/Web/Events/beforeunload
        window.addEventListener('beforeunload', function (event) {
            return event.returnValue = message;
        });
    },
    autoStart: true
};
