// jupyterlab/lib/renderers/index.js
// Copyright (c) Jupyter Development Team.
// Distributed under the terms of the Modified BSD License.
"use strict";
var widget_1 = require('./widget');
/**
 * A renderer for raw html.
 */
var HTMLRenderer = (function () {
    function HTMLRenderer() {
        /**
         * The mimetypes this renderer accepts.
         */
        this.mimetypes = ['text/html'];
    }
    /**
     * Whether the input can safely sanitized for a given mimetype.
     */
    HTMLRenderer.prototype.isSanitizable = function (mimetype) {
        return this.mimetypes.indexOf(mimetype) !== -1;
    };
    /**
     * Whether the input is safe without sanitization.
     */
    HTMLRenderer.prototype.isSafe = function (mimetype) {
        return false;
    };
    /**
     * Render the transformed mime bundle.
     */
    HTMLRenderer.prototype.render = function (options) {
        return new widget_1.RenderedHTML(options);
    };
    return HTMLRenderer;
}());
exports.HTMLRenderer = HTMLRenderer;
/**
 * A renderer for `<img>` data.
 */
var ImageRenderer = (function () {
    function ImageRenderer() {
        /**
         * The mimetypes this renderer accepts.
         */
        this.mimetypes = ['image/png', 'image/jpeg', 'image/gif'];
    }
    /**
     * Whether the input can safely sanitized for a given mimetype.
     */
    ImageRenderer.prototype.isSanitizable = function (mimetype) {
        return false;
    };
    /**
     * Whether the input is safe without sanitization.
     */
    ImageRenderer.prototype.isSafe = function (mimetype) {
        return this.mimetypes.indexOf(mimetype) !== -1;
    };
    /**
     * Render the transformed mime bundle.
     */
    ImageRenderer.prototype.render = function (options) {
        return new widget_1.RenderedImage(options);
    };
    return ImageRenderer;
}());
exports.ImageRenderer = ImageRenderer;
/**
 * A renderer for plain text and Jupyter console text data.
 */
var TextRenderer = (function () {
    function TextRenderer() {
        /**
         * The mimetypes this renderer accepts.
         */
        this.mimetypes = ['text/plain', 'application/vnd.jupyter.console-text'];
    }
    /**
     * Whether the input can safely sanitized for a given mimetype.
     */
    TextRenderer.prototype.isSanitizable = function (mimetype) {
        return false;
    };
    /**
     * Whether the input is safe without sanitization.
     */
    TextRenderer.prototype.isSafe = function (mimetype) {
        return this.mimetypes.indexOf(mimetype) !== -1;
    };
    /**
     * Render the transformed mime bundle.
     */
    TextRenderer.prototype.render = function (options) {
        return new widget_1.RenderedText(options);
    };
    return TextRenderer;
}());
exports.TextRenderer = TextRenderer;
/**
 * A renderer for raw `<script>` data.
 */
var JavascriptRenderer = (function () {
    function JavascriptRenderer() {
        /**
         * The mimetypes this renderer accepts.
         */
        this.mimetypes = ['text/javascript', 'application/javascript'];
    }
    /**
     * Whether the input can safely sanitized for a given mimetype.
     */
    JavascriptRenderer.prototype.isSanitizable = function (mimetype) {
        return false;
    };
    /**
     * Whether the input is safe without sanitization.
     */
    JavascriptRenderer.prototype.isSafe = function (mimetype) {
        return false;
    };
    /**
     * Render the transformed mime bundle.
     */
    JavascriptRenderer.prototype.render = function (options) {
        return new widget_1.RenderedJavascript(options);
    };
    return JavascriptRenderer;
}());
exports.JavascriptRenderer = JavascriptRenderer;
/**
 * A renderer for `<svg>` data.
 */
var SVGRenderer = (function () {
    function SVGRenderer() {
        /**
         * The mimetypes this renderer accepts.
         */
        this.mimetypes = ['image/svg+xml'];
    }
    /**
     * Whether the input can safely sanitized for a given mimetype.
     */
    SVGRenderer.prototype.isSanitizable = function (mimetype) {
        return false;
    };
    /**
     * Whether the input is safe without sanitization.
     */
    SVGRenderer.prototype.isSafe = function (mimetype) {
        return false;
    };
    /**
     * Render the transformed mime bundle.
     */
    SVGRenderer.prototype.render = function (options) {
        return new widget_1.RenderedSVG(options);
    };
    return SVGRenderer;
}());
exports.SVGRenderer = SVGRenderer;
/**
 * A renderer for PDF data.
 */
var PDFRenderer = (function () {
    function PDFRenderer() {
        /**
         * The mimetypes this renderer accepts.
         */
        this.mimetypes = ['application/pdf'];
    }
    /**
     * Whether the input can safely sanitized for a given mimetype.
     */
    PDFRenderer.prototype.isSanitizable = function (mimetype) {
        return false;
    };
    /**
     * Whether the input is safe without sanitization.
     */
    PDFRenderer.prototype.isSafe = function (mimetype) {
        return false;
    };
    /**
     * Render the transformed mime bundle.
     */
    PDFRenderer.prototype.render = function (options) {
        return new widget_1.RenderedPDF(options);
    };
    return PDFRenderer;
}());
exports.PDFRenderer = PDFRenderer;
/**
 * A renderer for LateX data.
 */
var LatexRenderer = (function () {
    function LatexRenderer() {
        /**
         * The mimetypes this renderer accepts.
         */
        this.mimetypes = ['text/latex'];
    }
    /**
     * Whether the input can safely sanitized for a given mimetype.
     */
    LatexRenderer.prototype.isSanitizable = function (mimetype) {
        return false;
    };
    /**
     * Whether the input is safe without sanitization.
     */
    LatexRenderer.prototype.isSafe = function (mimetype) {
        return this.mimetypes.indexOf(mimetype) !== -1;
    };
    /**
     * Render the mime bundle.
     */
    LatexRenderer.prototype.render = function (options) {
        return new widget_1.RenderedLatex(options);
    };
    return LatexRenderer;
}());
exports.LatexRenderer = LatexRenderer;
/**
 * A renderer for Jupyter Markdown data.
 */
var MarkdownRenderer = (function () {
    function MarkdownRenderer() {
        /**
         * The mimetypes this renderer accepts.
         */
        this.mimetypes = ['text/markdown'];
    }
    /**
     * Whether the input can safely sanitized for a given mimetype.
     */
    MarkdownRenderer.prototype.isSanitizable = function (mimetype) {
        return this.mimetypes.indexOf(mimetype) !== -1;
    };
    /**
     * Whether the input is safe without sanitization.
     */
    MarkdownRenderer.prototype.isSafe = function (mimetype) {
        return false;
    };
    /**
     * Render the mime bundle.
     */
    MarkdownRenderer.prototype.render = function (options) {
        return new widget_1.RenderedMarkdown(options);
    };
    return MarkdownRenderer;
}());
exports.MarkdownRenderer = MarkdownRenderer;
