// jupyterlab/lib/running/plugin.js
// Copyright (c) Jupyter Development Team.
// Distributed under the terms of the Modified BSD License.
"use strict";
var services_1 = require('../services');
var index_1 = require('./index');
/**
 * The default running sessions extension.
 */
exports.runningSessionsExtension = {
    id: 'jupyter.extensions.running-sessions',
    requires: [services_1.IServiceManager],
    activate: activateRunningSessions,
    autoStart: true
};
function activateRunningSessions(app, services) {
    var running = new index_1.RunningSessions({ manager: services });
    running.id = 'jp-running-sessions';
    running.title.label = 'Running';
    running.sessionOpenRequested.connect(function (sender, model) {
        var path = model.notebook.path;
        var name = path.split('/').pop();
        if (index_1.CONSOLE_REGEX.test(name)) {
            app.commands.execute('console:open', { id: model.id });
        }
        else {
            app.commands.execute('file-operations:open', { path: path });
        }
    });
    running.terminalOpenRequested.connect(function (sender, model) {
        app.commands.execute('terminal:open', { name: model.name });
    });
    // Rank has been chosen somewhat arbitrarily to give priority to the running
    // sessions widget in the sidebar.
    app.shell.addToLeftArea(running, { rank: 50 });
}
