// jupyterlab/lib/completer/handler.js
// Copyright (c) Jupyter Development Team.
// Distributed under the terms of the Modified BSD License.
"use strict";
/**
 * A completer handler for cell widgets.
 */
var CellCompleterHandler = (function () {
    /**
     * Construct a new completer handler for a widget.
     */
    function CellCompleterHandler(options) {
        this._activeCell = null;
        this._completer = null;
        this._kernel = null;
        this._pending = 0;
        this._completer = options.completer;
        this._completer.selected.connect(this.onCompletionSelected, this);
        this._completer.visibilityChanged.connect(this.onVisibilityChanged, this);
        this._kernel = options.kernel || null;
    }
    Object.defineProperty(CellCompleterHandler.prototype, "kernel", {
        /**
         * The kernel used by the completer handler.
         */
        get: function () {
            return this._kernel;
        },
        set: function (value) {
            this._kernel = value;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CellCompleterHandler.prototype, "activeCell", {
        /**
         * The cell widget used by the completer handler.
         */
        get: function () {
            return this._activeCell;
        },
        set: function (newValue) {
            if (newValue === this._activeCell) {
                return;
            }
            if (this._activeCell && !this._activeCell.isDisposed) {
                var editor = this._activeCell.editor;
                editor.textChanged.disconnect(this.onTextChanged, this);
                editor.completionRequested.disconnect(this.onCompletionRequested, this);
            }
            this._activeCell = newValue;
            if (this._activeCell) {
                var editor = this._activeCell.editor;
                editor.textChanged.connect(this.onTextChanged, this);
                editor.completionRequested.connect(this.onCompletionRequested, this);
            }
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(CellCompleterHandler.prototype, "isDisposed", {
        /**
         * Get whether the completer handler is disposed.
         *
         * #### Notes
         * This is a read-only property.
         */
        get: function () {
            return this._completer === null;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Dispose of the resources used by the handler.
     */
    CellCompleterHandler.prototype.dispose = function () {
        if (this.isDisposed) {
            return;
        }
        this._completer = null;
        this._kernel = null;
        this._activeCell = null;
    };
    /**
     * Make a complete request using the kernel.
     */
    CellCompleterHandler.prototype.makeRequest = function (request) {
        var _this = this;
        if (!this._kernel) {
            return Promise.reject(new Error('no kernel for completion request'));
        }
        var content = {
            code: request.currentValue,
            cursor_pos: request.position
        };
        var pending = ++this._pending;
        return this._kernel.requestComplete(content).then(function (msg) {
            _this.onReply(pending, request, msg);
        });
    };
    /**
     * Receive a completion reply from the kernel.
     */
    CellCompleterHandler.prototype.onReply = function (pending, request, msg) {
        // If we have been disposed, bail.
        if (this.isDisposed) {
            return;
        }
        // If a newer completion request has created a pending request, bail.
        if (pending !== this._pending) {
            return;
        }
        var value = msg.content;
        var model = this._completer.model;
        if (!model) {
            return;
        }
        // Completion request failures or negative results fail silently.
        if (value.status !== 'ok') {
            model.reset();
            return;
        }
        // Update the original request.
        model.original = request;
        // Update the options.
        model.options = value.matches;
        // Update the cursor.
        model.cursor = { start: value.cursor_start, end: value.cursor_end };
    };
    /**
     * Handle a text changed signal from an editor.
     */
    CellCompleterHandler.prototype.onTextChanged = function (editor, change) {
        if (!this._completer.model) {
            return;
        }
        this._completer.model.handleTextChange(change);
    };
    /**
     * Handle a completion requested signal from an editor.
     */
    CellCompleterHandler.prototype.onCompletionRequested = function (editor, request) {
        if (!this._kernel || !this._completer.model) {
            return;
        }
        this.makeRequest(request);
    };
    /**
     * Handle a visiblity change signal from a completer widget.
     */
    CellCompleterHandler.prototype.onVisibilityChanged = function (completer) {
        if (completer.isDisposed || completer.isHidden) {
            if (this._activeCell) {
                this._activeCell.activate();
            }
        }
    };
    /**
     * Handle a completion selected signal from the completion widget.
     */
    CellCompleterHandler.prototype.onCompletionSelected = function (widget, value) {
        if (!this._activeCell || !this._completer.model) {
            return;
        }
        var patch = this._completer.model.createPatch(value);
        if (!patch) {
            return;
        }
        var cell = this._activeCell;
        cell.model.source = patch.text;
        cell.editor.setCursorPosition(patch.position);
    };
    return CellCompleterHandler;
}());
exports.CellCompleterHandler = CellCompleterHandler;
