// jupyterlab/lib/common/observablestring.js
// Copyright (c) Jupyter Development Team.
// Distributed under the terms of the Modified BSD License.
"use strict";
var signaling_1 = require('phosphor/lib/core/signaling');
/**
 * A concrete implementation of [[IObservableString]]
 */
var ObservableString = (function () {
    /**
     * Construct a new observable string.
     */
    function ObservableString(initialText) {
        if (initialText === void 0) { initialText = ''; }
        this._text = '';
        this._isDisposed = false;
        this.text = initialText;
    }
    Object.defineProperty(ObservableString.prototype, "text", {
        /**
         * Get the value of the string.
         */
        get: function () {
            return this._text;
        },
        /**
         * Set the value of the string.
         */
        set: function (value) {
            this._text = value;
            this.changed.emit({
                type: 'set',
                start: 0,
                end: value.length,
                value: value
            });
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Insert a substring.
     *
     * @param index - The starting index.
     *
     * @param text - The substring to insert.
     */
    ObservableString.prototype.insert = function (index, text) {
        this._text = this._text.slice(0, index) +
            text +
            this._text.slice(index);
        this.changed.emit({
            type: 'insert',
            start: index,
            end: index + text.length,
            value: text
        });
    };
    /**
     * Remove a substring.
     *
     * @param start - The starting index.
     *
     * @param end - The ending index.
     */
    ObservableString.prototype.remove = function (start, end) {
        var oldValue = this._text.slice(start, end);
        this._text = this._text.slice(0, start) +
            this._text.slice(end);
        this.changed.emit({
            type: 'remove',
            start: start,
            end: end,
            value: oldValue
        });
    };
    /**
     * Set the ObservableString to an empty string.
     */
    ObservableString.prototype.clear = function () {
        this.text = '';
    };
    Object.defineProperty(ObservableString.prototype, "isDisposed", {
        /**
         * Test whether the string has been disposed.
         */
        get: function () {
            return this._isDisposed;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Dispose of the resources held by the string.
     */
    ObservableString.prototype.dispose = function () {
        if (this._isDisposed) {
            return;
        }
        this._isDisposed = true;
        this.clear();
        signaling_1.clearSignalData(this);
    };
    return ObservableString;
}());
exports.ObservableString = ObservableString;
// Define the signals for the `ObservableString` class.
signaling_1.defineSignal(ObservableString.prototype, 'changed');
