// jupyterlab/lib/imagewidget/widget.js
// Copyright (c) Jupyter Development Team.
// Distributed under the terms of the Modified BSD License.
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var widget_1 = require('phosphor/lib/ui/widget');
var docregistry_1 = require('../docregistry');
/**
 * The class name added to a imagewidget.
 */
var IMAGE_CLASS = 'jp-ImageWidget';
/**
 * A widget for images.
 */
var ImageWidget = (function (_super) {
    __extends(ImageWidget, _super);
    /**
     * Construct a new image widget.
     */
    function ImageWidget(context) {
        var _this = this;
        _super.call(this, { node: Private.createNode() });
        this._scale = 1;
        this._context = context;
        this.node.tabIndex = -1;
        this.addClass(IMAGE_CLASS);
        if (context.model.toString()) {
            this.update();
        }
        context.pathChanged.connect(function () { return _this.update(); });
        context.model.contentChanged.connect(function () { return _this.update(); });
        context.fileChanged.connect(function () { return _this.update(); });
    }
    Object.defineProperty(ImageWidget.prototype, "scale", {
        /**
         * The scale factor for the image.
         */
        get: function () {
            return this._scale;
        },
        set: function (value) {
            if (value === this._scale) {
                return;
            }
            this._scale = value;
            var scaleNode = this.node.querySelector('div');
            var transform;
            transform = "scale(" + value + ")";
            scaleNode.style.transform = transform;
            this.update();
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Dispose of the resources used by the widget.
     */
    ImageWidget.prototype.dispose = function () {
        if (this.isDisposed) {
            return;
        }
        this._context = null;
        _super.prototype.dispose.call(this);
    };
    /**
     * Handle `update-request` messages for the widget.
     */
    ImageWidget.prototype.onUpdateRequest = function (msg) {
        this.title.label = this._context.path.split('/').pop();
        var cm = this._context.contentsModel;
        if (cm === null) {
            return;
        }
        var content = this._context.model.toString();
        var src = "data:" + cm.mimetype + ";" + cm.format + "," + content;
        this.node.querySelector('img').setAttribute('src', src);
    };
    /**
     * Handle `'activate-request'` messages.
     */
    ImageWidget.prototype.onActivateRequest = function (msg) {
        this.node.focus();
    };
    return ImageWidget;
}(widget_1.Widget));
exports.ImageWidget = ImageWidget;
/**
 * A widget factory for images.
 */
var ImageWidgetFactory = (function (_super) {
    __extends(ImageWidgetFactory, _super);
    function ImageWidgetFactory() {
        _super.apply(this, arguments);
    }
    /**
     * Create a new widget given a context.
     */
    ImageWidgetFactory.prototype.createNewWidget = function (context) {
        return new ImageWidget(context);
    };
    return ImageWidgetFactory;
}(docregistry_1.ABCWidgetFactory));
exports.ImageWidgetFactory = ImageWidgetFactory;
/**
 * A namespace for image widget private data.
 */
var Private;
(function (Private) {
    /**
     * Create the node for the image widget.
     */
    function createNode() {
        var node = document.createElement('div');
        var innerNode = document.createElement('div');
        var image = document.createElement('img');
        node.appendChild(innerNode);
        innerNode.appendChild(image);
        return node;
    }
    Private.createNode = createNode;
})(Private || (Private = {}));
