// jupyterlab/lib/about/plugin.js
// Copyright (c) Jupyter Development Team.
// Distributed under the terms of the Modified BSD License.
"use strict";
var commandpalette_1 = require('../commandpalette');
var _1 = require('./');
/**
 * The about page extension.
 */
exports.aboutExtension = {
    id: 'jupyter.extensions.about',
    activate: activateAbout,
    autoStart: true,
    requires: [commandpalette_1.ICommandPalette]
};
function activateAbout(app, palette) {
    var model = new _1.AboutModel();
    var widget = new _1.AboutWidget();
    widget.model = model;
    widget.id = 'about-jupyterlab';
    widget.title.label = 'About';
    widget.title.closable = true;
    widget.node.style.overflowY = 'auto';
    var command = 'about-jupyterlab:show';
    app.commands.addCommand(command, {
        label: 'About JupyterLab',
        execute: function () {
            if (!widget.isAttached) {
                app.shell.addToMainArea(widget);
            }
            app.shell.activateMain(widget.id);
        }
    });
    palette.addItem({ command: command, category: 'Help' });
}
