// jupyterlab/lib/markdownwidget/plugin.js
// Copyright (c) Jupyter Development Team.
// Distributed under the terms of the Modified BSD License.
"use strict";
var docregistry_1 = require('../docregistry');
var rendermime_1 = require('../rendermime');
var widget_1 = require('./widget');
/**
 * The class name for all main area portrait tab icons.
 */
var PORTRAIT_ICON_CLASS = 'jp-MainAreaPortraitIcon';
/**
 * The class name for the text editor icon from the default theme.
 */
var TEXTEDITOR_ICON_CLASS = 'jp-ImageTextEditor';
/**
 * The markdown handler extension.
 */
exports.markdownHandlerExtension = {
    id: 'jupyter.extensions.rendered-markdown',
    requires: [docregistry_1.IDocumentRegistry, rendermime_1.IRenderMime],
    activate: function (app, registry, rendermime) {
        var factory = new widget_1.MarkdownWidgetFactory({
            name: 'Rendered Markdown',
            fileExtensions: ['.md'],
            rendermime: rendermime
        });
        var icon = PORTRAIT_ICON_CLASS + " " + TEXTEDITOR_ICON_CLASS;
        factory.widgetCreated.connect(function (sender, widget) {
            widget.title.icon = icon;
        });
        registry.addWidgetFactory(factory);
    },
    autoStart: true
};
